src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1006,21 +1009,50 @@ public SolutionCompilationState WithOptions(SolutionOptionSet options)
             this.SolutionState.WithOptions(options));
     }
 
+    public Task<SolutionCompilationState> WithFrozenPartialCompilationsAsync(CancellationToken cancellationToken)
+        => _cachedFrozenSnapshot.GetValueAsync(cancellationToken);
+
+    private SolutionCompilationState ComputeFrozenSnapshot(CancellationToken cancellationToken)
+    {
+        var newIdToProjectStateMap = this.SolutionState.ProjectStates;
+        var newIdToTrackerMap = _projectIdToTrackerMap;
+
+        foreach (var (projectId, projectState) in newIdToProjectStateMap)
+        {
+            // if we don't have one or it is stale, create a new partial solution
+            var tracker = GetCompilationTracker(projectId);
+            var newTracker = tracker.FreezePartialState(this, cancellationToken);
+
+            Contract.ThrowIfFalse(newIdToProjectStateMap.ContainsKey(projectId));
+            newIdToProjectStateMap = newIdToProjectStateMap.SetItem(projectId, newTracker.ProjectState);
+            newIdToTrackerMap = newIdToTrackerMap.SetItem(projectId, newTracker);
+        }
+
+        var newState = this.SolutionState.Branch(
+            idToProjectStateMap: newIdToProjectStateMap,
+            dependencyGraph: SolutionState.CreateDependencyGraph(this.SolutionState.ProjectIds, newIdToProjectStateMap));
+        var newCompilationState = this.Branch(
+            newState,
+            newIdToTrackerMap);
+
+        // Set the frozen solution to be its own frozen solution.  Freezing multiple times is a no-op.
+        newCompilationState._cachedFrozenSnapshot = _cachedFrozenSnapshot;
+
+        return newCompilationState;
+    }
+
     /// <summary>
-    /// Creates a branch of the solution that has its compilations frozen in whatever state they are in at the time, assuming a background compiler is
-    /// busy building this compilations.
-    ///
-    /// A compilation for the project containing the specified document id will be guaranteed to exist with at least the syntax tree for the document.
-    ///
+    /// Creates a branch of the solution that has its compilations frozen in whatever state they are in at the time,
+    /// assuming a background compiler is busy building this compilations.
+    /// <para/>
+    /// A compilation for the project containing the specified document id will be guaranteed to exist with at least the
+    /// syntax tree for the document.
+    /// <para/>
     /// This not intended to be the public API, use Document.WithFrozenPartialSemantics() instead.
     /// </summary>
     public SolutionCompilationState WithFrozenPartialCompilationIncludingSpecificDocument(
         DocumentId documentId, CancellationToken cancellationToken)
     {
-        // in progress solutions are disabled for some testing
-        if (this.Services.GetService<IWorkspacePartialSolutionsTestHook>()?.IsPartialSolutionDisabled == true)
-            return this;
```

Comment:
>return this; [](http://example.com/codeflow?start=12&length=12)

Trying to understand WithFrozenPartialCompilationIncludingSpecificDocumentWorker below a bit better:

Is there a reason that the code in that method outside the if condition can't go inside it?

Reply:
i wanted to make sure we didn't access this.SolutionState or `this`.  We're first overwriting `currentCompilationState` N times.  We then pass that final SolutionCompilationSatate into the final static helper.  This helps ensure we don't accidentally use the wrong state object (or access anything off of 'this').

Reply:
I worded my question poorly, will clarify in person tomorrow.